File: WorkflowPatternFinder/WorkflowEventLogFixer/WorkflowEventLogFixer/WorkflowEventLogFixer/SubTreeFinder.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace WorkflowEventLogFixer
{
    public static class SubTreeFinder
    {
        public static string _pythonExe;
        public static string _word2VecQueryFile = Path.Combine(Directory.GetCurrentDirectory(), "Scripts", "query.py");
        public static string _word2VecTrainedModelPath;

        public static bool IsValidSubTree(ProcessTree tree, ProcessTree pattern, bool induced)
        {
            var tNode = tree.GetRoot();
            var pNode = pattern.GetRoot();
            return DoesBranchContainPattern(tNode, pNode, induced);
        }

        private static bool DoesBranchContainPattern(Node tNode, Node pNode, bool induced)
        {
            // if current node in tree = current node in pattern
            if(AreSimilar(tNode, pNode))
            //if(tNode.GetEvent() == pNode.GetEvent())
            {
                var pChildren = pNode.GetChildren();
                var tChildren = tNode.GetChildren();

                // check the siblings-condition. All siblings in the pattern should be included in the tree.
                if(!ContainsSiblings(tNode, pNode))
                {
                    return false;
                }

                // continue searching for children of the current pattern node.
                if(pChildren.Any())
                {
                    foreach(var patternChild in pChildren)
                    {
                        // For every tree child, check whether this child can be mapped to a pattern child.
                        foreach(Node treeChild in tChildren)
                        {
                            var patternChildFound = DoesBranchContainPattern(treeChild, patternChild, induced);
                            if(patternChildFound)
                            {
                                return true;
                            }
                        }
                        break;
                    }
                }

                // in case the current pattern node is not a parent, then we can say that the subtree of pNode as root occurs in the given tree.
                else
                {
                    return true;
                }
            }

            // return false if the current pattern searched is already partly found, but this node does not correspond to the pattern's structure.
            else if(induced && !pNode.IsRoot())
            {
                return false;
            }

            // continue searching for the pattern node. 
            // in the case of induced subtrees, this can only be done when the pNode is the root.
            // in the case of embedded subtrees, also internal nodes may be searched.
            if(induced && pNode.IsRoot() || !induced)
            {
                foreach(Node treeChild in tNode.GetChildren())
                {
                    var found = DoesBranchContainPattern(treeChild, pNode, induced);
                    if(found)
                    {
                        return true;
                    }
                }
            }

            // return false when none of the above searches have succeeded.
            return false;
        }

        private static bool AreSimilar(Node tNode, Node pNode)
        {
            if(pNode.GetType() == tNode.GetType())
            {
                if(pNode.GetType() == ProcessTreeLoader.NodeType.manualTask)
                {
                    return AreSimilarAccordingToDoc2Vec(tNode, pNode);
                }
                return string.Equals(tNode.GetEvent(), tNode.GetEvent(), StringComparison.CurrentCultureIgnoreCase);
            }
            return false;
        }

        private static bool AreSimilarAccordingToDoc2Vec(Node tNode, Node pNode)
        {
            var treeSentence = tNode.GetEvent().Replace('|', ' ').ToLower();
            var patternSentence = pNode.GetEvent().Replace('|', ' ').ToLower();
            ProcessStartInfo start = new ProcessStartInfo
            {
                FileName = _pythonExe,
                Arguments = $"{_word2VecQueryFile} \"{_word2VecTrainedModelPath}\" \"{treeSentence}\" \"{patternSentence}\"",
                UseShellExecute = false,
                RedirectStandardOutput = true,
                CreateNoWindow = true
            };
            //cmd is full path to python.exe
            //args is path to .py file and any cmd line args
            using(Process process = Process.Start(start))
            {
                using(StreamReader reader = process?.StandardOutput)
                {
                    string result = reader?.ReadToEnd();
                    Console.Write(result);
                    if(!string.IsNullOrEmpty(result) && double.TryParse(result.Substring(0, 3).Replace('.', ','), out double number))
                    {
                        if(number > 0.8)
                        {
                            return true;
                        }
                    }
                    else if(string.IsNullOrEmpty(result))
                    {
                        return false;
                    }

                }
            }
            return false;
        }

        private static bool ContainsSiblings(Node tNode, Node pNode)
        {
            if(tNode.GetSiblings().Intersect(pNode.GetSiblings()).ToList().Count == pNode.GetSiblings().Count)
            {
                return true;
            }
            return false;
        }

        public static void SetPythonExe(string pythonExe)
        {
            _pythonExe = pythonExe;
        }

        public static void SetTrainedModelPath(string modelPath)
        {
            _word2VecTrainedModelPath = modelPath;
        }
    }
}

File: WorkflowPatternFinder/Gensim/SubTreeFinder.py
import gensim # learn more: https://python.org/pypi/gensim
import sys
import enum
from ProcessTree import *
from Node import *
from ProcessTreeLoader import *
from Query import *

class SubTreeFinder(object):
  _pythonExe = None
  _word2VecTrainedModelPath = None
  _query = None
  
  def LoadWord2VecModel(self):
    if self._word2VecTrainedModelPath is not None:
      self._query = Query()
      self._query.LoadModel(self._word2VecTrainedModelPath)
    else:
      raise ValueError('The path to model is not set yet.')

  def IsValidSubTree(self, tree, pattern, induced=False):
    # print('induced is ',induced)
    tNode = tree.GetRoot()
    pNode = pattern.GetRoot()
    return self.DoesBranchContainPattern(tNode, pNode, [], induced)

  def GetValidSubTrees(self, tree, pattern, induced=False):
    flatten = lambda x: [item for sublist in x for item in sublist]
    pNode = pattern.GetRoot()
    treeNodes = tree.GetNodes()
    allPatterns = []
    for tNode in treeNodes:
      if tNode.GetId() not in [i[0] for i in allPatterns]:
        result = self.DoesBranchContainPattern(tNode, pNode, allPatterns, induced)
        if result[0]:
          newPattern = True
          for member in result[2]:
            print("matching: member = ", str(member), "result[2] = ", str(result[2]))
            if member in allPatterns:
              newPattern = False
              break
          if newPattern:
            for tuple in result[2]:
              allPatterns.append(tuple)    
    return allPatterns

  def DoesBranchContainPattern(self, tNode, pNode, allPatterns, induced=False):
    patternMembers = []
    simTuple = self.AreSimilar(tNode, pNode)
    if self.IsTupleTrue(simTuple):
      patternMembers.append((tNode.GetId(), self.GetPatternWord(simTuple)))
      print ('pattern members: '+str(patternMembers))
      pChildren = pNode.GetChildren()
      tChildren = tNode.GetChildren()
      if self.ContainsSiblings(tNode, pNode):
        if(any(pChildren)):
          pFound = []
          treeIds = []
          for patternChild in pChildren:
            for treeChild in tChildren:
              if treeChild.GetId() not in treeIds:
                patternChildFound = self.DoesBranchContainPattern(treeChild, patternChild, allPatterns, induced)
                if self.IsTupleTrue(patternChildFound):
                    pFound.append((self.GetScore(patternChildFound), self.GetPatternMembers(patternChildFound)))
                    treeIds.append(treeChild.GetId())
                    break
          if len(pFound) == len(pChildren):
              for node in pFound:
                print('node: ', str(node), ", patternMembers: ", str(patternMembers)) 
                patternMembers.extend((node[1]))
              pFound.append((self.GetScore(simTuple), self.GetPatternMembers(simTuple)))
              total = pFound
              min_score = 1
              try:
                for found in total:
                  score = found[0]
                  if score < min_score:
                    min_score = score
                return (True, min_score, patternMembers)
              except:
                raise ValueError('object in total is not correct')
        else:
          print ('in else statement')
          return (True, self.GetScore(simTuple), patternMembers)
    if induced and pNode.IsRoot() or not induced:
      for treeChild in tNode.GetChildren():
        found = self.DoesBranchContainPattern(treeChild, pNode, allPatterns, induced)
        if(self.IsTupleTrue(found)):
          return (True, self.GetScore(found), self.GetPatternMembers(found), self.GetPatternWord(found))
    return (False, 0)

  def IsTupleTrue(self, tuple):
    try:
      return tuple[0]
    except:
      raise ValueError("This is not a proper tuple.")

  def GetScore(self, tuple):
    try:
      return tuple[1]
    except:
      raise ValueError("This is not a proper tuple.")

  def GetPatternMembers(self, tuple):
    try:
      return tuple[2]
    except:
      raise ValueError("This is not a proper tuple.")

  def GetPatternWord(self, tuple):
    try:
      return tuple[3]
    except:
      return "";

  def AreSimilar(self, tNode, pNode):
    if pNode.GetType() == tNode.GetType():
      if pNode.GetType() == 'ManualTask':
        return self.AreSimilarAccordingToDoc2Vec(tNode, pNode)
      else:
        return (True, 1, tNode.GetId(), '')
    else:
        return (False, 0, tNode.GetId(), '')
    
  def AreSimilarAccordingToDoc2Vec(self, tNode, pNode):
    score = self._query.GetSimilarity(tNode.GetEvent(), pNode.GetEvent())
    if score[0] > self._simThreshold:
      return (True, score[0], tNode.GetId(), score[1])
    return (False, score[0], tNode.GetId(), score[1])
    
  def ContainsSiblings(self, tNode, pNode):
    similars = []
    pSiblings = pNode.GetSiblings()
    tSiblings = tNode.GetSiblings()
    for pSibling in pSiblings:
      for tSibling in tSiblings:
        if self.IsTupleTrue(self.AreSimilar(pSibling, tSibling)):
          similars.append(tSibling.GetId())
    if len(similars) == len(pSiblings):
      return (True, similars)
    return (False, [])
  
  def SetTrainedModelPath(self, modelPath):
    self._word2VecTrainedModelPath = modelPath
    self.LoadWord2VecModel()
    
  def SetSimilarityThreshold(self, simThreshold):
    self._simThreshold = float(simThreshold)
    # print('sim threshold: ' + str(self._simThreshold))